proposals/plugins.md

```diff
@@ -0,0 +1,146 @@
+
+# Initial Use Cases
+
+1. A user wishes to deploy a component by default to their cluster instances outside of Kubernetes, e.g. a 
+monitoring agent.
+1. A user wishes to deploy a component into their cluster by default that is not part of core
+Kubernetes such as cluster wide services, addons, ingress controller or even persistent volumes.
+Reasons can vary from company restrictions to always having this component form part of every 
+cluster. e.g. [the rescheduler](https://github.com/kubernetes-incubator/kube-aws/issues/118)
+1. A user wishes to add [customised parts to the CloudFormation template](https://github.com/kubernetes-incubator/kube-aws/issues/96#issuecomment-263835808) 
+produced by kube-aws. Anything they add would need access to the same parameters and template 
+engine that already exists.
+1. A user wishes to customise existing parts of the CloudFormation template produced by kube-aws
+without necessarily having to diff all the files for changes with there sometimes being large
+conflicts or huge updates due to kube-aws changes that can be difficult to digest while
+being able to merge in the simple changes they need. Examples 
+[here](https://github.com/kubernetes-incubator/kube-aws/issues/340) and 
+[here](https://github.com/kubernetes-incubator/kube-aws/issues/338)
+
+# Specific Examples
+
+* Deployment outside of Kubernetes - Dex uses some additional API server parameter, Datadog agent still has more functionality if not using the containerised version
+* Deployment to Kubernetes - rescheduler, kube2iam
+* Add new parts to CloudFormation - bastion host
+* Customisations to existing parts of CloudFormation - IAM roles/groups
+
+# Plugin Principles
+
+1. Plugin functionality should first be added in places where kube-aws needs it most, i.e. where
+kube-aws receives more feature requests which do not fit with the core feature set.
+1. kube-aws should dogfood the plugins system, i.e. use the plugins system where possible on existing functionality
+1. Everything that is pluggable should be kept as modular as possible, from lists of IAM roles to kube API server parameters.
+1. Where possible, checks will be added to kube-aws to ensure the integrity of the cluster via validation or disallowing particular changes. e.g. disallow overriding node pool init metadata
+   * Note, this could be gradually added rather than forming part of the first implementation
+   * Some of this will likely be provided by CloudFormation but kube-aws could possibly replicas the checks to save time for the stack create/update to fail
+
+# Design
+
+## Plugin Types
+
+* Built-in plugins versus ad-hoc user specific plugins
+  * Some plugins will be provided by kube-aws and some will be specific to a user
+  * Both will use the same mechanisms to hook into kube-aws/deploy etc
+  * Both types of plugins will be enabled and configured in the same way
+* Built-in plugins versus contrib plugins
```

Comment:
Please let me also add that *contrib* plugins is hosted remotely in the kube-aws repo which can be installed locally and then work similarly to *adhoc* plugins.

Related: https://github.com/kubernetes-incubator/kube-aws/pull/751/files#r126319557